OpenRA.Mods.Common/Widgets/ObserverArmyIconsWidget .cs

```diff
@@ -0,0 +1,253 @@
+				Ui.Root.Get<TooltipContainerWidget>(TooltipContainer));
+
+			Init();
+
+			actorDataCache = new Cache<string, ActorData>(LoadActorDataCache);
+		}
+
+		ActorData LoadActorDataCache(string name)
+		{
+			var actorInfo = unitLookup[name].Item1;
+			var bi = actorInfo.TraitInfo<BuildableInfo>();
+
+			var rsi = actorInfo.TraitInfo<RenderSpritesInfo>();
+
+			return new ActorData(bi, rsi);
+		}
+
+		protected ObserverArmyIconsWidget(ObserverArmyIconsWidget other)
+			: base(other)
+		{
+			GetPlayer = other.GetPlayer;
+			world = other.world;
+			worldRenderer = other.worldRenderer;
+
+			IconWidth = other.IconWidth;
+			IconHeight = other.IconHeight;
+			IconSpacing = other.IconSpacing;
+			iconSize = new float2(IconWidth, IconHeight);
+
+			MinWidth = other.MinWidth;
+
+			TooltipUnit = other.TooltipUnit;
+			GetTooltipUnit = () => TooltipUnit;
+
+			TooltipTemplate = other.TooltipTemplate;
+			TooltipContainer = other.TooltipContainer;
+
+			tooltipContainer = Exts.Lazy(() =>
+				Ui.Root.Get<TooltipContainerWidget>(TooltipContainer));
+
+			Init();
+
+			actorDataCache = new Cache<string, ActorData>(LoadActorDataCache);
+		}
+
+		void Init()
+		{
+			var buildable = world.Map.Rules.Actors.Values
+				.Where(x => x.HasTraitInfo<BuildableInfo>())
+				.SelectMany(x => x.TraitInfo<BuildableInfo>().Queue, (ai, q) => new Tuple<string, ActorInfo, BuildableInfo>(q, ai, ai.TraitInfo<BuildableInfo>()));
+
+			var pqueue = world.Map.Rules.Actors.Values
+				.SelectMany(a => a.TraitInfos<ProductionQueueInfo>());
+
+			unitLookup = buildable.Join(pqueue, tuple => tuple.Item1, pq => pq.Type, (tuple, info) => new { tuple.Item2.Name, Ai = tuple.Item2, tuple.Item3.BuildPaletteOrder, info.DisplayOrder })
+				.Distinct()
+				.ToDictionary(t => t.Name, o => new Tuple<ActorInfo, int, int>(o.Ai, o.DisplayOrder, o.BuildPaletteOrder));
+		}
+
+		public override void Draw()
+		{
+			armyUnits.Clear();
+			armyIconsBounds.Clear();
+
+			var player = GetPlayer();
+			if (player == null)
+				return;
+
+			var playerStatistics = stats.Update(player);
+
+			var items = playerStatistics.Units
+				.Where(i => i.Value.Count > 0)
+				.OrderBy(i => unitLookup[i.Key].Item2)
+				.ThenBy(i => unitLookup[i.Key].Item3)
+				.ToList();
+
+			Bounds.Width = items.Count * (IconWidth + IconSpacing);
+
+			var queueCol = 0;
+			foreach (var unit in items)
+			{
+				var name = unit.Key;
+				var actorInfo = unitLookup[name].Item1;
+
+				var actorData = actorDataCache[name];
+
+				var rsi = actorData.Rsi;
+				var icon = new Animation(world, rsi.GetImage(actorInfo, world.Map.Rules.Sequences, player.Faction.Name));
+
+				icon.Play(actorData.BuildableInfo.Icon);
+				var topLeftOffset = new float2(queueCol * (IconWidth + IconSpacing), 0);
+
+				var iconTopLeft = RenderOrigin + topLeftOffset;
+				var centerPosition = iconTopLeft;
+
+				WidgetUtils.DrawSHPCentered(icon.Image, centerPosition + 0.5f * iconSize, worldRenderer.Palette(actorData.BuildableInfo.IconPalette), 0.5f);
```

Comment:
Can we please rebase on #17581 and follow the approach from 0a9a04981516c3e93d0d58a8d5280e9a806f6683 to improve the icon downscaling quality?